IA-885 & IA-901 restrict actions submissions perm

quang-le opened the pull request:
## Changes

- changed "permission" field to "permissions" in `menuItems`. It now takes an array i.o a string
- gave access to Forms to "iaso_submissions" permission
- updated `ProtectedRoutes` to check for an array of permissions i.o single one
- users with `iaso_submissions` permission can view `/forms/list`, but write actions are not available
- id. for ArchivedForms: the actions column is removed

https://user-images.githubusercontent.com/38907762/140543052-6c7570ce-3355-4995-b42e-6ab17af2bd1c.mov

beygorghor commented on hat/assets/js/apps/Iaso/domains/forms/index.js:
```diff
@@ -41,8 +44,9 @@ const Forms = ({ params, showOnlyDeleted }) => {
               null,
-              true,
-              true,
+              userHasFormsPermission,
```
lol, maybe use an object here...

quang-le replied:
lol

quang-le replied:
by changing this I found a bug elsewhere 😅

olethanh commented:
![image](https://user-images.githubusercontent.com/82500/140959212-0134840e-2a27-4bca-b411-bfbafcf70ff4.png)
we should probably hide or disable the create button for Form.

olethanh commented on iaso/api/forms.py:
```diff
@@ -21,6 +21,18 @@ def has_permission(self, request, view):
         return request.user.is_authenticated and request.user.has_perm("menupermissions.iaso_forms")
 
 
+# Used for forms, which have limited access front-end side with submissions permission
+class HasFormOrSubmissionPermission(permissions.BasePermission):
+    def has_permission(self, request, view):
+        if request.method in permissions.SAFE_METHODS:
+            return True
+
+        return request.user.is_authenticated and (
+            request.user.has_perm("menupermissions.iaso_forms")
+            or request.user.has_perm("menupermissions.iaso_submissions")
+        )
+
+
```
not sure it make sense to me why an use with the submission permission can modify form?

quang-le commented:
## Update

- removed create button when no "forms" permission
- removed `iaso_submissions`'s write access to forms api
- fixed bug with archived forms (displayed action buttons only to superusers)
- updated and added test

quang-le commented:
## Update

Fixed IA-901 here to avoid git gymnastics:

- added handleFormDelete method to Org Units details

olethanh commented:
The "Show submission button" in actions need to be removed if the user don't have the submission permissions
![image](https://user-images.githubusercontent.com/82500/141776781-ed5022e1-c395-4b9d-8e53-1d545b8bdaa3.png)
Or we say that the user who has form also has implicitly the Submission perm but then we need to correct the perm on the Submission page, and to make it more clear for the user

And inversely in submissions page, we should remove the link to the Form if we don't have the form permission
![image](https://user-images.githubusercontent.com/82500/141776527-1152d41d-a886-47c6-8e7b-56d73dad8bf6.png)

quang-le commented:
> The "Show submission button" in actions need to be removed if the user don't have the submission permissions ![image](https://user-images.githubusercontent.com/82500/141776781-ed5022e1-c395-4b9d-8e53-1d545b8bdaa3.png) Or we say that the user who has form also has implicitly the Submission perm but then we need to correct the perm on the Submission page, and to make it more clear for the user
> 
> And inversely in submissions page, we should remove the link to the Form if we don't have the form permission ![image](https://user-images.githubusercontent.com/82500/141776527-1152d41d-a886-47c6-8e7b-56d73dad8bf6.png)

You should be able to see the form with the submissions permission, just not edit them, so I think the link should stay.

I will have a look at the rest

quang-le commented:
## Update 16/11/2021

- remove "see submissions" action button when submissions permissions is absent
- hide "forms" tab from org unit details when forms permission is absent
- make "forms" tab from org unit details behave like forms page in other cases (superuser, forms permission with and without submissions permission)
- update instances tests

https://user-images.githubusercontent.com/38907762/141987054-88261319-69ad-400d-8625-c536d098c6a0.mov

olethanh commented:
The link from the orgunit tab fail because there is a levels at the end? 
See :  http://localhost:8081/dashboard/forms/submissions/formIds/1/tab/list/columns/updated_at,org_unit__name,created_at,status/levels/29679

beygorghor reviewed (approved):
Nice job, thanks for the tests
